persistence/data_models_json_fields.go

```diff
@@ -91,6 +91,164 @@ func (s *StateExecutionSequenceMapsJson) CompleteNewStateExecution(stateId strin
+//
+//	state_1, 1: (q1, 2),
+//	state_1, 2: (q2, 3),
+func (s *StateExecutionWaitingQueuesJson) Consume(message LocalQueueMessageInfoJson) (*string, []uuid.UUID) {
+	s.UnconsumedMessageQueueDedupIdsMap[message.QueueName] = append(s.UnconsumedMessageQueueDedupIdsMap[message.QueueName], message.DedupId)
+
+	for stateExecutionIdKey, commands := range s.StateToCommandsMap {
+		for i, command := range commands {
+			if command.GetQueueName() != message.QueueName || int(command.GetCount()) > len(s.UnconsumedMessageQueueDedupIdsMap[message.QueueName]) {
+				continue
+			}
+
+			consumedDedupIds := s.UnconsumedMessageQueueDedupIdsMap[message.QueueName][:int(command.GetCount())]
+
+			s.UnconsumedMessageQueueDedupIdsMap[message.QueueName] =
+				s.UnconsumedMessageQueueDedupIdsMap[message.QueueName][int(command.GetCount()):]
+			if len(s.UnconsumedMessageQueueDedupIdsMap[message.QueueName]) == 0 {
+				delete(s.UnconsumedMessageQueueDedupIdsMap, message.QueueName)
+			}
+
+			s.StateToCommandsMap[stateExecutionIdKey] = append(s.StateToCommandsMap[stateExecutionIdKey][:i],
+				s.StateToCommandsMap[stateExecutionIdKey][i+1:]...)
+			if len(s.StateToCommandsMap[stateExecutionIdKey]) == 0 {
+				delete(s.StateToCommandsMap, stateExecutionIdKey)
+			}
+
+			return &stateExecutionIdKey, consumedDedupIds
+		}
+	}
+
+	return nil, []uuid.UUID{}
+}
+
+// ConsumeFor return a bool indicating if the stateExecutionId can complete the local queue commands, and an array of all the consumed dedupIds.
+//
+// E.g., given UnconsumedMessageQueueCountMap as:
+//
+// (q1, 2), (q2, 2), (q3, 1)
+//
+// and StateToCommandsMap[stateExecutionId] as:
+//
+// (q1, 1), (q2, 2)
+//
+// and isAllOfCompletion as true. Then after ConsumeFor, the UnconsumedMessageQueueCountMap becomes:
+//
+// (q1, 1), (q3, 1)
+//
+// and returns:
+//
+// (true, [q1_dedup_id_1, q2_dedup_id_1, q2_dedup_id_2])
+func (s *StateExecutionWaitingQueuesJson) ConsumeFor(stateExecutionId StateExecutionId, isAllOfCompletion bool) (bool, []uuid.UUID) {
+	stateExecutionIdKey := stateExecutionId.GetStateExecutionId()
+
+	remainingCommands := []xdbapi.LocalQueueCommand{}
+	consumedDedupIds := []uuid.UUID{}
+
+	idx := 0
+
+	for i, command := range s.StateToCommandsMap[stateExecutionIdKey] {
+		idx = i
+
+		dedupIds, ok := s.UnconsumedMessageQueueDedupIdsMap[command.GetQueueName()]
+
+		if !ok || int(command.GetCount()) > len(dedupIds) {
+			remainingCommands = append(remainingCommands, command)
+			continue
+		}
+
+		consumedDedupIds = append(consumedDedupIds, s.UnconsumedMessageQueueDedupIdsMap[command.GetQueueName()][:int(command.GetCount())]...)
+
+		s.UnconsumedMessageQueueDedupIdsMap[command.GetQueueName()] = s.UnconsumedMessageQueueDedupIdsMap[command.GetQueueName()][int(command.GetCount()):]
+
+		if len(s.UnconsumedMessageQueueDedupIdsMap[command.GetQueueName()]) == 0 {
+			delete(s.UnconsumedMessageQueueDedupIdsMap, command.GetQueueName())
```

Comment:
Still going through the details, but I wonder if we can just keep this method read-only, and the caller will call `CleanupFor` instead for deleting

Reply:
This len check for UnconsumedMessageQueueDedupIdsMap needs to be performed everytime a consumption happen, otherwise the data will contain many empty `{}`.

Reply:
Okay, I think I see what you meant. 

I think then it should be `ConsumeWithCheckingLocalQueueWaitingComplete`, instead of `CheckCanCompleteLocalQueueWaiting` (sorry for the back and forth). 

But I still a little confused why we have this and the `Consume` method above.

Reply:
> But I still a little confused why we have this and the Consume method above.

To maintain the map as `{}` but not `{{}, {}, {}}`?

Reply:
Yeah I got it now. I probably should remove this comment

Reply:
When reviewing code, I used a review session to keep accumulate comments  and publish all at the end. Some of the earlier comments became stale that I forgot to remove before publishing